test/plugin/base_test.rb

```diff
@@ -626,6 +654,118 @@ def test_struct_payload_json_container_log
     end
   end
 
+  # Verify that we drop the log entries when 'require_valid_tags' is true and
+  # any non-string tags or tags with non-utf8 characters are detected.
+  def test_reject_invalid_tags_with_require_valid_tags_true
+    setup_gce_metadata_stubs
+    invalid_tags = NON_STRING_TAGS.merge(EMPTY_TAG).merge(NON_UTF8_TAGS)
+    invalid_tags.each do |tag, _|
+      setup_logging_stubs do
+        @logs_sent = []
+        d = create_driver(REQUIRE_VALID_TAGS_CONFIG, tag)
+        d.emit('msg' => log_entry(0))
+        d.run
+      end
+      verify_log_entries(0, COMPUTE_PARAMS, 'structPayload')
+    end
+  end
+
+  # Verify that tags are properly encoded. When 'require_valid_tags' is true, we
+  # only accept string tags with utf8 characters.
+  def test_encode_tags_with_require_valid_tags_true
+    setup_gce_metadata_stubs
+    VALID_TAGS.each do |tag, encoded_tag|
+      setup_logging_stubs([COMPUTE_PARAMS.merge(log_name: encoded_tag)]) do
+        @logs_sent = []
+        d = create_driver(REQUIRE_VALID_TAGS_CONFIG, tag)
+        d.emit('msg' => log_entry(0))
+        d.run
+      end
+      verify_log_entries(1, COMPUTE_PARAMS, 'structPayload')
+      # Each batch in the logs sent shares the same log name, so we only verify
+      # the first one.
+      assert_equal "projects/#{PROJECT_ID}/logs/#{encoded_tag}",
+                   @logs_sent[0]['logName']
+    end
+  end
+
+  # Verify that tags extracted from container names are properly encoded.
+  def test_encode_tags_from_container_name_with_require_valid_tags_true
```

Comment:
I tried to include that test initially, then figured it was kind of redundant:

For empty string case:
The tag fails the kubernetes regex matching, thus fall back to the generic tag as the log name case, which is already covered by `test_reject_invalid_tags_with_require_valid_tags_true`. In other words, it's not really testing container specific stuff.

For non-string cases:
By the time it reaches to the point where the kubernetes regex can be matched and container name being extracted, the tag is certainly gonna be a string.

For non-utf8 cases:
With `require_valid_tags` true, any tag containing non-utf8 strings would be rejected early on before reaching any container specific logic, which is also covered by `test_reject_invalid_tags_with_require_valid_tags_true` already.

Thoughts?